### PR TITLE
coprocessor/dag/expr: add some un-pushed builtin UDFs(BitCount)

### PR DESCRIPTION
**Pull request by spongedu**
## What have you changed? (mandatory)

implement builtin function: `BitCount`  in `coprocessor`.
This is for #3275

## What are the type of the changes? (mandatory)

Improvement

## How has this PR been tested? (mandatory)

unittest

## Refer to a related PR or issue link (optional)

#3275

### CODE REVIEW

**Comment by TennyZhuang** [comment]
It seems that you can use https://doc.rust-lang.org/std/primitive.u64.html#method.count_ones directly？

**Comment by TennyZhuang** [comment]
your implementation is very inefficient, which take O(bit) = 64 to calculate the ret. There are some algorithm which only take O(1-bit). And there exists built-in count_ones.

**Comment by spongedu** [comment]
@TennyZhuang it seems so. thx, I'll refine this place.

File: src/coprocessor/dag/expr/builtin_other.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+use std::i64;
+
+use super::{EvalContext, Result, ScalarFunc};
+use coprocessor::codec::Datum;
+
+impl ScalarFunc {
+    #[inline]
+    pub fn bit_count(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let arg = try_opt!(self.children[0].eval_int(ctx, row));
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_other.rs]: what if `eval_int` returns an overflow error? (https://github.com/pingcap/tidb/blob/master/expression/builtin_other.go#L723)

**Comment by hicqu** [comment]
LGTM.

File: src/coprocessor/dag/expr/builtin_other.rs (hunk 1 above)
Comment (hicqu) [reply src/coprocessor/dag/expr/builtin_other.rs]: mysql returns `63`.

File: src/coprocessor/dag/expr/builtin_other.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    #[inline]
+    pub fn bit_count(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let res = self.children[0].eval_int(ctx, row);
+        match res {
+            Ok(r) => if let Some(v) = r {
+                Ok(Some(i64::from(v.count_ones())))
+            } else {
+                Ok(None)
+            },
+            Err(e) => if e.is_overflow() {
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_other.rs]: Any test cover it?

Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_other.rs]: @AndreMouche How do we construct an overflow case here ?

Comment (AndreMouche) [reply src/coprocessor/dag/expr/builtin_other.rs]: Maybe we can construct the argument with `CastStrAsInt` which may throw an overflow.

Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_other.rs]: OK

File: src/coprocessor/dag/expr/builtin_other.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                Ok(Some(64i64))
```
Comment (siddontang) [inline src/coprocessor/dag/expr/builtin_other.rs]: using 64 is enough here?

Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_other.rs]: @siddontang  this is corresponding to the implementation in [TiDB](https://github.com/pingcap/tidb/blob/master/expression/builtin_other.go)

File: src/coprocessor/dag/expr/builtin_other.rs (hunk 2 above)
Comment (Hoverbear) [reply src/coprocessor/dag/expr/builtin_other.rs]: Any ypdate?

Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_other.rs]: Add overflow case. PTAL @Hoverbear @AndreMouche  :)

File: src/coprocessor/dag/expr/builtin_other.rs (hunk 3 above)
Comment (siddontang) [reply src/coprocessor/dag/expr/builtin_other.rs]: what I mean is `Ok(Some(64))`

Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_other.rs]: oh, yes, will refine here. :)

File: src/coprocessor/dag/expr/builtin_other.rs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    pub fn bit_count(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let res = self.children[0].eval_int(ctx, row);
+        match res {
+            Ok(r) => if let Some(v) = r {
+                Ok(Some(i64::from(v.count_ones())))
+            } else {
+                Ok(None)
+            },
+            Err(e) => if e.is_overflow() {
+                Ok(Some(64))
```
Comment (breeswish) [inline src/coprocessor/dag/expr/builtin_other.rs]: I would like to see test cases in some edge cases. In MySQL:

```
mysql> select bit_count(9223372036854775807);
+--------------------------------+
| bit_count(9223372036854775807) |
+--------------------------------+
|                             63 |
+--------------------------------+
1 row in set (0.00 sec)

mysql> select bit_count(9223372036854775808);
+--------------------------------+
| bit_count(9223372036854775808) |
+--------------------------------+
|                              1 |
+--------------------------------+
1 row in set (0.00 sec)

mysql> select bit_count(9223372036854775809);
+--------------------------------+
| bit_count(9223372036854775809) |
+--------------------------------+
|                              2 |
+--------------------------------+
1 row in set (0.00 sec)

mysql> select bit_count(111111111111111111111111111111111111111111111111111111111111111);
+----------------------------------------------------------------------------+
| bit_count(111111111111111111111111111111111111111111111111111111111111111) |
+----------------------------------------------------------------------------+
|                                                                         63 |
+----------------------------------------------------------------------------+
1 row in set, 1 warning (0.01 sec)

mysql> select bit_count('111111111111111111111111111111111111111111111111111111111111111');
+------------------------------------------------------------------------------+
| bit_count('111111111111111111111111111111111111111111111111111111111111111') |
+------------------------------------------------------------------------------+
|                                                                           64 |
+------------------------------------------------------------------------------+
1 row in set, 1 warning (0.01 sec)

mysql> select bit_count(11111111112222222222);
+---------------------------------+
| bit_count(11111111112222222222) |
+---------------------------------+
|                              37 |
+---------------------------------+
1 row in set (0.03 sec)

mysql> select bit_count('11111111112222222222');
+-----------------------------------+
| bit_count('11111111112222222222') |
+-----------------------------------+
|                                37 |
+-----------------------------------+
1 row in set (0.00 sec)
```

**Comment by AndreMouche** [comment]
Any update?

**Comment by spongedu** [comment]
@AndreMouche  In Progress. Will update later :)

File: src/coprocessor/dag/expr/builtin_other.rs (hunk 4 above)
Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_other.rs]: @breeswish  the above test cases added. PTAL :)

**Comment by spongedu** [comment]
@AndreMouche @hicqu  Add more test cases, PTAL :)

**Comment by huachaohuang** [comment]
@breeswish PTAL

**Pull request merged**